[RDPF-127] 패스워드, 아이디 찾기 api 작성

kma952727 opened the pull request:
## Changes ✏️

1. 1번 api : "/api/send-password-to-mail", 2번 api : "/api/find-username" api 작성
2. 1번 api -  email, username을 받아 연결된 email로 패스워드 정보를 보냅니다.
3. 2번 api - email은 받아 username 중간부터 끝까지 "*"로 암호화된 "username"을 응답합니다.
## Test checklist 🧪

<!-- 기능 추가 시 작성한 테스트 코드 목록을 작성해주세요. -->

- [테스트코드 추가하겠습니다. ]

kma952727 committed:
[RDPF-127] 패스워드, 아이디 찾기 api 작성

mooowu commented on perfume-api/src/main/java/io/perfume/api/user/adapter/in/http/CommonController.java:
```diff
@@ -0,0 +1,37 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/v1")
+public class CommonController {
```
`common`보다는 좀 더 의미가 있는 이름이 사용되면 좋을 것 같은데 어떠신가요..?!

kma952727 replied:
좋습니다!

mooowu commented on perfume-api/src/main/java/io/perfume/api/user/adapter/in/http/CommonController.java:
```diff
@@ -0,0 +1,37 @@
+    private final FindUserUseCase findUserUseCase;
+
+    // TODO OAUTH 가입자의 경우 uesrname도 같이 넣는 흐름이 이상하다.
+    @GetMapping("/send-password-to-mail")
```
url path가 너무 서술적이라 개인적으로 조금 이질감이 있는 것 같아요! 관련해서 [이 글](https://learn.microsoft.com/ko-kr/azure/architecture/best-practices/api-design)이 좋은 것 같은데 참고해보셔도 좋을 것 같습니당 👍 !

kma952727 replied:
너무 좋은데요? 이런 것 찾고있었는데 감사합니다 ㅎ.ㅎ

mooowu commented on perfume-api/src/main/java/io/perfume/api/user/application/service/FindUserService.java:
```diff
@@ -23,4 +27,25 @@ private UserResult toDto(User user) {
+  public void sendPasswordToEmail(String email, String username) {
+
+    User user = userQueryRepository.findOneByEmailAndUsername(email, username).orElseThrow(NotFoundUserException::new);
+    mailSender.send(email, "[read a perfume] 비밀번호 찾기 ", "[read a perfume] 사용자의 암호 : " + user.getPassword());
```
`user.getPassword()` 결과는 해쉬값이라 비밀번호 원문을 알기는 어려울 것 같아요!; 아마 이 과정은 **비밀번호 재설정**으로 가는게 좋을 것 같네용 🤔

kma952727 replied:
피그마에 보니깐 비밀번호 찾기로 보이길래 복호화 안될텐데 생각했는데 꼼꼼하게 체크하겠습니다 ㅜ.ㅜ;

mooowu replied:
@kma952727 아하 그렇네요 😓 요것도 나중에 디자이너님이랑 같이 논의 해보면 좋을 것 같군용 👍 !

mooowu commented on perfume-api/src/main/java/io/perfume/api/user/application/service/FindUserService.java:
```diff
@@ -23,4 +27,25 @@ private UserResult toDto(User user) {
+
+  @Override
+  public String getEncryptedUsernameByEmail(String email) {
+
+    User user = userQueryRepository.findOneByEmail(email).orElseThrow(NotFoundUserException::new);
+    String names = user.getUsername();
+
+    // encrypt 알고리즘
+    int length = names.length();
+
+    int halfLength = length / 2;
+
+    return names.substring(0, halfLength) + "*".repeat(length - halfLength);
```
이 부분은 User domain 내부로 옮겨도 좋을 것 같네요 👍 !

mooowu commented on perfume-api/src/main/java/io/perfume/api/user/adapter/out/persistence/UserQueryPersistenceAdapter.java:
```diff
@@ -17,30 +17,40 @@ public class UserQueryPersistenceAdapter implements UserQueryRepository {
+    UserJpaEntity userJpaEntity = jpaQueryFactory.selectFrom(QUserJpaEntity.userJpaEntity)
+            .where(
+                    QUserJpaEntity.userJpaEntity.username.eq(username)
+                            .and(QUserJpaEntity.userJpaEntity.email.eq(email)))
```
`.and(QUserJpaEntity.userJpaEntity.deletedAt.isNull()))` 조건이 추가되면 좋을 것 같아용 😃 !

kma952727 replied:
내일 추가하겠습니다 ㅎ..ㅎ;

mooowu reviewed (commented):
작업 고생하셨습니다 👍 👍 !! 몇가지 의견 남겨용 ㅎ!

kma952727 committed:
[RDPF-127] 사용자 이름을 암호화하는 로직 캡슐화

1. 사용자 아이디 찾기 기능에 사용되는 로직에서 User Doamin의 내부구현이 노출되는 현상이 발생하여 해당 정보의 전문가가 수정하도록 개선
2. 사용자의 서비스 이용에 부가적인 도움을 주는 컨트롤러명 변경

kma952727 committed:
refactor : 패스워드를 분실한 사용자가 사용하는 기능, 구조 변경

1. 패스워드를 분실한 사용자에게 해싱된 현재 암호를 이메일로 보내지 않고 랜덤한 값으로 변경된 새로운 암호를 이메일로 전송
2. 암호 찾기 관련 플로우 기존 "Find"라는 이름으로 작성된 기능(service, usecase)에서 분리
3. User domain에 임의의 새로운 암호를 할당하는 기능 추가
4.

kma952727 committed:
Merge branch 'develop' into RDPF-127-Backend-api-find-my-info-api

mooowu reviewed (approved):
고생하셨습니다 👍 !

oliviarla commented:
@kma952727 
리뷰 진행중인데, 미리 반영되어야 할 게 있습니다!

Sample쪽에 UserNotFoundException이 있어서 해당 클래스는 user/application/exception으로 옮기도록 하겠습니다.
그리고 원래 있던 곳에는 SampleNotFoundException 추가하겠씁니다!!
즉, NotFoundUserException 새로 만들 필요 없이 UserNotFoundException를 사용하면 됩니당!

kma952727 committed:
refactor : findOneByEmailAndUsername() 에 SoftDelete 체크 조건 추가

kma952727 committed:
Merge branch 'develop' into RDPF-127-Backend-api-find-my-info-api

kma952727 committed:
refactor : 이메일을 사용하여 암호화된 유저네임 조회 유즈케이스 로직 수정

1. User Model이 가지고 있는 "자신의 유저네임을 암호화해서 반환"하는 메서드 "getEncryptedUsernameByEmail()" 이름을 "getEncryptedUsername()"으로 변경
2. merge conflict 해결하다가 실수로 "이메일을 사용하여 암호화된 유저네임 조회 유즈케이스"에 사용되는 port, service layer method를 삭제함, 새로 기능 작성 ( + FindUserUseCase에서 관련 기능 분리, 따로 파일만들어서 사용)
3.

kma952727 committed:
test commit  : 커밋 내역이 푸쉬가 안된다.
